### PR TITLE
Introduce StashCollection.Apply and Pop

### PR DESCRIPTION
**Pull request by Therzok**
Supersedes #871

### CODE REVIEW

**Comment by Therzok** [comment]
cc @nulltoken

File: LibGit2Sharp/StashApplyOptions.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        Default = 0,
+
+        /// <summary>
+        /// Reinstate both index and working directory stashed
+        /// changes.
+        /// </summary>
+        ReinstateIndex = (1 << 0),
+    }
+}
+
```
Comment (nulltoken) [inline LibGit2Sharp/StashApplyOptions.cs]: The `E_TOOMANYNEWLINES` police has been alerted :trolleybus:

File: LibGit2Sharp/StashApplyProgress.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        /// The modified files are being written to disk.
+        /// </summary>
+        CheckoutModified,
+
+        /// <summary>
+        /// The stash was applied successfully.
+        /// </summary>
+        Done,
+    }
+
```
Comment (nulltoken) [inline LibGit2Sharp/StashApplyProgress.cs]: :police_car:

File: LibGit2Sharp/StashApplyProgress.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+}
+
```
Comment (nulltoken) [inline LibGit2Sharp/StashApplyProgress.cs]: :cop:

File: LibGit2Sharp/StashApplyStatus.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        /// </summary>
+        Conflicts,
+
+        /// <summary>
+        /// The stash index given was not found.
+        /// </summary>
+        NotFound,
+    }
+}
+
```
Comment (nulltoken) [inline LibGit2Sharp/StashApplyStatus.cs]: :oncoming_police_car:

File: LibGit2Sharp/StashCollection.cs [hunk 5]
```diff
@@ -132,6 +132,82 @@ public virtual Stash Add(Signature stasher, string message, StashModifiers optio
...
+            }
+        }
+
+        /// <summary>
+        /// Pops a single stashed state from the stash list using the default options.
+        /// </summary>
+        /// <param name="index">the index of the stash to remove (0 being the most recent one).</param>
+        public virtual StashApplyStatus Pop(int index)
+        {
+            return Pop(index, null);
```
Comment (nulltoken) [inline LibGit2Sharp/StashCollection.cs]: How about passing a new instance of `StashApplyOptions` and ensuring `public virtual StashApplyStatus Pop(int index, StashApplyOptions options)` guards against `null` options?

File: LibGit2Sharp/StashCollection.cs [hunk 6]
```diff
@@ -132,6 +132,82 @@ public virtual Stash Add(Signature stasher, string message, StashModifiers optio
...
+        {
+            return Apply(index, null);
+        }
+
+        /// <summary>
+        /// Pops a single stashed state from the stash list
+        /// </summary>
+        /// <param name="index">the index of the stash to remove (0 being the most recent one).</param>
+        /// <param name="options">the options to use for checking out the stash.</param>
+        public virtual StashApplyStatus Pop(int index, StashApplyOptions options)
```
Comment (nulltoken) [inline LibGit2Sharp/StashCollection.cs]: Shouldn't we ensure that index > 0?

File: LibGit2Sharp/StashCollection.cs [hunk 7]
```diff
@@ -132,6 +132,82 @@ public virtual Stash Add(Signature stasher, string message, StashModifiers optio
...
+            }
+        }
+
+        /// <summary>
+        /// Applies a single stashed state from the stash list using the default options.
+        /// </summary>
+        /// <param name="index">the index of the stash to remove (0 being the most recent one).</param>
+        public StashApplyStatus Apply(int index)
+        {
+            return Apply(index, null);
```
Comment (nulltoken) [inline LibGit2Sharp/StashCollection.cs]: Same comment regarding `null` usage here

File: LibGit2Sharp/StashApplyProgress.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+{
+    /// <summary>
+    /// The current progress of the stash application.
+    /// </summary>
+    public enum StashApplyProgress
+    {
+        /// <summary>
+        /// The stash application has just started.
+        /// </summary>
+        None = 0,
```
Comment (nulltoken) [inline LibGit2Sharp/StashApplyProgress.cs]: `Initiated`?

File: LibGit2Sharp/StashCollection.cs (hunk 5 above)
Comment (Therzok) [inline LibGit2Sharp/StashCollection.cs]: Hm, that's different from the behaviour of other options, iirc.

https://github.com/libgit2/libgit2sharp/blob/vNext/LibGit2Sharp/Repository.cs#L1189

File: LibGit2Sharp/StashCollection.cs (hunk 6 above)
Comment (Therzok) [inline LibGit2Sharp/StashCollection.cs]: I think libgit2 gives E_NOTFOUND on that. I could do an argument check.

File: LibGit2Sharp/StashApplyProgress.cs (hunk 8 above)
Comment (Therzok) [inline LibGit2Sharp/StashApplyProgress.cs]: Actually, I'm not even sure that its called the first time. Judging from https://github.com/libgit2/libgit2/pull/3018/files#diff-f54e4a7883cb063f36ebd8f4206ce55fR290, I may have misnamed it.

File: LibGit2Sharp/StashCollection.cs (hunk 5 above)
Comment (nulltoken) [inline LibGit2Sharp/StashCollection.cs]: Hmmm. You're right. Let's keep it that way. I'm going to open an issue about this.

File: LibGit2Sharp.Tests/StashFixture.cs [hunk 9]
```diff
@@ -204,6 +204,152 @@ public void CanStashIgnoredFiles()
...
+        }
+
+        [Fact]
+        public void CanStashAndPop()
+        {
+            string path = SandboxStandardTestRepo();
+            using (var repo = new Repository(path))
+            {
+                var stasher = Constants.Signature;
+
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/StashFixture.cs]: `Assert.Equal(0, repo.Stashes.Count());`

File: LibGit2Sharp.Tests/StashFixture.cs [hunk 10]
```diff
@@ -204,6 +204,152 @@ public void CanStashIgnoredFiles()
...
+                const string filename = "staged_file_path.txt";
+                const string contents = "I'm staged\n";
+                string contentsNew = "I'm staged" + Environment.NewLine;
+                Touch(repo.Info.WorkingDirectory, filename, contents);
+                repo.Stage(filename);
+
+                repo.Stashes.Add(stasher, "This stash with default options");
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/StashFixture.cs]: `Assert.Equal(1, repo.Stashes.Count());`

File: LibGit2Sharp/StashApplyOptions.cs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    /// <summary>
+    /// The flags which control whether the index should be reinstated.
+    /// </summary>
+    [Flags]
+    public enum StashApplyModifiers
+    {
+        /// <summary>
+        /// Default. Reinstate working directory stashed
```
Comment (nulltoken) [inline LibGit2Sharp/StashApplyOptions.cs]: Default. Reinstate stashed working directory

File: LibGit2Sharp/StashApplyOptions.cs [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        /// changes.
+        /// </summary>
+        Default = 0,
+
+        /// <summary>
+        /// Reinstate both index and working directory stashed
```
Comment (nulltoken) [inline LibGit2Sharp/StashApplyOptions.cs]: Reinstate both stashed index and working directory

File: LibGit2Sharp/StashApplyOptions.cs (hunk 11 above)
Comment (nulltoken) [inline LibGit2Sharp/StashApplyOptions.cs]: Drop this. I overlooked the line below

File: LibGit2Sharp/StashApplyOptions.cs (hunk 12 above)
Comment (nulltoken) [inline LibGit2Sharp/StashApplyOptions.cs]: Drop this. I overlooked the line below

**Comment by nulltoken** [comment]
Although AppVeyor and Travis both disagree with me, I :heart: this very much!!

Very neat PR, Sir!

File: LibGit2Sharp/StashApplyOptions.cs (hunk 11 above)
Comment (Therzok) [inline LibGit2Sharp/StashApplyOptions.cs]: `stashed` is a typo, it should be `stash`.

**Comment by Therzok** [comment]
Test coverage could be better, I agree, but nothing much I can do about that atm. CI failed because they http 503'd while getting nugets.

**Comment by nulltoken** [comment]
>CI failed because they http 503'd while getting nugets

AppVeyor cringes with the following

```
LibGit2Sharp.Tests.StashFixture.CanStashAndPop [FAIL]
   Assert.Equal() Failure
   Position: First difference is at position 10
   Expected: I'm staged
             
   Actual:   I'm staged
   Stack Trace:
      c:\projects\libgit2sharp\LibGit2Sharp.Tests\StashFixture.cs(259,0): at LibGit2Sharp.Tests.StashFixture.CanStashAndPop()
 
LibGit2Sharp.Tests.MetaFixture.TypesInLibGit2SharpMustBeExtensibleInATestingContext [FAIL]
   
   'LibGit2Sharp.StashCollection' cannot be easily abstracted in a testing context. Please make sure it either has a public constructor, or an empty protected constructor.
       - Method 'LibGit2Sharp.StashApplyStatus Apply(Int32)' must be virtual
   Stack Trace:
      c:\projects\libgit2sharp\LibGit2Sharp.Tests\MetaFixture.cs(112,0): at LibGit2Sharp.Tests.MetaFixture.TypesInLibGit2SharpMustBeExtensibleInATestingContext()
 ```

**Comment by Therzok** [comment]
@nulltoken was thinking of making a common function for both Apply and Pop, so it basically is like `StashApplyStatus Common(int index, StashApplyOptions options, Func<RepositorySafeHandle, int, GitStashApplyOpts, StashApplyStatus)` and we call as:

`Common(index, options, Proxy.git_stash_apply)` and `Common(index, options, Proxy.git_stash_pop` from the functions and do the common code in there. I find this whole mess as or a delegate with that prototype. Solves us the mess of maintaining two pieces of code. :dango:

File: LibGit2Sharp/StashCollection.cs (hunk 6 above)
Comment (Therzok) [inline LibGit2Sharp/StashCollection.cs]: Fixed.

File: LibGit2Sharp.Tests/StashFixture.cs (hunk 9 above)
Comment (Therzok) [inline LibGit2Sharp.Tests/StashFixture.cs]: Fixt.

**Comment by nulltoken** [comment]
>Solves us the mess of maintaining two pieces of code.

:+1:

**Comment by Therzok** [comment]
Done!

File: LibGit2Sharp/StashApplyOptions.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    /// <summary>
+    /// The flags which control whether the index should be reinstated.
+    /// </summary>
+    [Flags]
+    public enum StashApplyModifiers
+    {
+        /// <summary>
+        /// Default. Reinstate working directory stashed changes.
+        /// </summary>
+        Default = 0,
```
Comment (nulltoken) [inline LibGit2Sharp/StashApplyOptions.cs]: How about making it explicit? eg. `ReinstateWorkingDirectory`

File: LibGit2Sharp/StashApplyOptions.cs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+        /// <summary>
+        /// Reinstate both index and working directory stashed changes.
+        /// </summary>
+        ReinstateIndex = (1 << 0),
```
Comment (nulltoken) [inline LibGit2Sharp/StashApplyOptions.cs]: And `ReinstateWorkingDirectoryAndIndex`

File: LibGit2Sharp/Core/GitStashApplyOpts.cs [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+        public uint Version = 1;
+
+        public StashApplyModifiers Flags;
+        public GitCheckoutOpts CheckoutOptions;
+
+        public stash_apply_progress_cb ApplyProgressCallback;
+        public IntPtr ProgressPayload;
+    }
+}
+
```
Comment (nulltoken) [inline LibGit2Sharp/Core/GitStashApplyOpts.cs]: E_TOOMANYNEWLINES!!1! :trolleybus:

File: LibGit2Sharp/StashCollection.cs (hunk 5 above)
Comment (nulltoken) [inline LibGit2Sharp/StashCollection.cs]: cf. #1076

File: LibGit2Sharp/StashApplyOptions.cs (hunk 14 above)
Comment (Therzok) [inline LibGit2Sharp/StashApplyOptions.cs]: The thing is, it's weird git naming. ReinstateIndex gives a merge conflict, default doesn't pop.

Comment (nulltoken) [inline LibGit2Sharp/StashApplyOptions.cs]: >The thing is, it's weird git naming.

Ok. 

>ReinstateIndex gives a merge conflict

Could you please update the xml doc to make that explicit to the reader.

>default doesn't pop.

Huh? I'm sorry. You kinda lost me here, mate.

pop-fart

Comment (Therzok) [inline LibGit2Sharp/StashApplyOptions.cs]: I was really bad at thinking when I responded. I'm bringing this libgit2 [comment](https://github.com/libgit2/libgit2/pull/3018#issuecomment-90225220) into place.

What I meant is that ReinstateIndex will not create an index when conflicts arise (doesn't pop/apply), but when the default is used, it'll generate an index with conflicts.

File: LibGit2Sharp/Core/GitStashApplyOpts.cs (hunk 15 above)
Comment (Therzok) [inline LibGit2Sharp/Core/GitStashApplyOpts.cs]: Will fix.

File: LibGit2Sharp/StashApplyOptions.cs (hunk 14 above)
Comment (nulltoken) [inline LibGit2Sharp/StashApplyOptions.cs]: Duh. Stash is hard. But your explanation makes it much clearer. Thanks!

 I *really* think this deserves to be described in the xml doc.

Comment (Therzok) [inline LibGit2Sharp/StashApplyOptions.cs]: Yeah so, naming should stay as is, so we match cli git and libgit2 flags.

I'll update the docs to reflect the stuff.

**Comment by Therzok** [comment]
@nulltoken done.

**Comment by nulltoken** [comment]
:100:

**Pull request merged**